gc/base/MemorySubSpaceSemiSpace.cpp

```diff
@@ -970,37 +973,49 @@ MM_MemorySubSpaceSemiSpace::checkSubSpaceMemoryPostCollectResize(MM_EnvironmentB
+					double desiredContractionFactor, adjustedContractionFactor;
 
-				if (desiredContractionFactor > extensions->dnssMaximumContraction) {
-					adjustedContractionFactor = extensions->dnssMaximumContraction;
-				} else if (desiredContractionFactor < extensions->dnssMinimumContraction) {
-					adjustedContractionFactor = extensions->dnssMinimumContraction;
-				} else {
-					adjustedContractionFactor = desiredContractionFactor;
-				}
+					/* Try to reach 200% of the expected minimum time ratio through contraction */
+					desiredContractionFactor = OMR_MIN(extensions->dnssExpectedTimeRatioMinimum  * 2, expectedTimeRatio);
+					desiredContractionFactor = desiredContractionFactor - _averageScavengeTimeRatio;
 
-				/* Adjust the average scavenge to interval ratio */
-				_averageScavengeTimeRatio += adjustedContractionFactor;
+					if (desiredContractionFactor > extensions->dnssMaximumContraction) {
+						adjustedContractionFactor = extensions->dnssMaximumContraction;
+					} else if (desiredContractionFactor < extensions->dnssMinimumContraction) {
+						adjustedContractionFactor = extensions->dnssMinimumContraction;
+					} else {
+						adjustedContractionFactor = desiredContractionFactor;
+					}
 
+					/* Adjust the average scavenge to interval ratio */
+					_averageScavengeTimeRatio += adjustedContractionFactor;
 
-				_contractionSize = MM_Math::roundToCeiling(extensions->heapAlignment, (uintptr_t)(getCurrentSize() * adjustedContractionFactor));
-				_contractionSize = MM_Math::roundToCeiling(regionSize, _contractionSize);
 
-				if(debug) {
-					omrtty_printf("\tContract decision - contractFactor desired: %lf adjusted: %lf size: %u\n", desiredContractionFactor, adjustedContractionFactor, _contractionSize);
-					omrtty_printf("\tContract decision - current size: %d contracted size: %d\n", getCurrentSize(), getCurrentSize() - _contractionSize);
-					omrtty_printf("\tContract decision - new time ratio:%lf\n\n\n", _averageScavengeTimeRatio);
-				}
+					_contractionSize = MM_Math::roundToCeiling(extensions->heapAlignment, (uintptr_t)(getCurrentSize() * adjustedContractionFactor));
+					_contractionSize = MM_Math::roundToCeiling(regionSize, _contractionSize);
+
+					if(debug) {
+						omrtty_printf("\tContract decision - contractFactor desired: %lf adjusted: %lf size: %u\n", desiredContractionFactor, adjustedContractionFactor, _contractionSize);
+						omrtty_printf("\tContract decision - current size: %d contracted size: %d\n", getCurrentSize(), getCurrentSize() - _contractionSize);
+						omrtty_printf("\tContract decision - new time ratio:%lf\n\n\n", _averageScavengeTimeRatio);
+					}
+
+					extensions->heap->getResizeStats()->setLastContractReason(SCAV_RATIO_TOO_LOW);
 
-				extensions->heap->getResizeStats()->setLastContractReason(SCAV_RATIO_TOO_LOW);
+				} else if ((0 != extensions->heap->getActualSoftMxSize(env, MEMORY_TYPE_NEW)) && (extensions->heap->getActualSoftMxSize(env, MEMORY_TYPE_NEW) < getCurrentSize())) {
```

Comment:
for slighly easier to read code you can put actualSoftMx into a local (used 3x), but not a must - up to you.

Reply:
Agreed, I'll move that to a local, before the surrounding if statement